_posts/2016-03-07-http2-and-you.md

```diff
@@ -0,0 +1,93 @@
+
+[http2](https://github.com/molnarg/node-http2) is a package available for Node.js I used to create a basic [HTTP/2 demo](https://github.com/adamhenson/http2-demo). With it - you can utilize the protocol with TLS on localhost. In the demo I setup an example of server push. I have a queue of files to push like so:
+
+```javascript
+const FILES = [
+  {
+    'headers' : {
+      'content-type' : 'text/css'
+    },
+    'path' : '/public/css/main.css'
+  },
+  {
+    'headers' : {
+      'content-type' : 'image/jpeg'
+    },
+    'path' : '/public/images/nyc.jpg'
+  }
+];
+```
+
+And below is the request callback function that loops through the file queue, creates a stream for each, and pushes to the server response.
+
+```javascript
+function onRequest(request, response) {
+  if(response.push) {
+    FILES.forEach((file, index) => {
+      let push = response.push(file.path);
+      push.writeHead(200, file.headers);
+      fs.createReadStream(path.join(__dirname, file.path)).pipe(push);
+      if(index === FILES.length - 1) response.end(Template.output(FILES));
+    });
+  }
+}
+```
+
+While logging (via bunyan), the output for the image push looks like the following:
+
+```bash
+19:29:19.398Z  INFO server/http: Promising push stream (e=1, s=12, method=GET, scheme=https, authority=localhost:8080, path=/public/images/nyc.jpg)
+19:29:19.398Z  INFO server/http: Sending server response (e=1, s=14, status=200)
+    headers: {
+      "content-type": "image/jpeg",
+      "date": "Sun, 06 Mar 2016 19:29:19 GMT"
+    }
+```
+
+Someone could write an article about the details of what's happening under the hood here. The server is actually pushing a promise, and then it sends the accompanying response. Grey areas exist in the subject of caching, and communication between client and server in the case of browser caching. In the protocol's current state, results could vary by implementation, but theoretically the browser cache should be honored. This answer from [Stack Overflow](http://stackoverflow.com/questions/29352282/do-browser-cancel-server-push-when-resource-is-in-cache) summarizes it pretty well - "When the client receives the PUSH_PROMISE, it can look at the URI, and figure out the cache status of this resource. Browsers typically use different caches for normally received resources and pushed resources. If the cache is still valid, the client may cancel the pushed stream by sending a RST_STREAM frame to the server for that stream." ([sbordet](http://stackoverflow.com/users/1215076/sbordet))
```

Comment:
> Someone could write an article about the details of what's happening under the hood here. The server is actually pushing a promise, and then it sends the accompanying response.

This bit isn't very clear to me, could you perhaps go into more detail?

Also I think the `Template.output(FILES)` call needs explaining too - what does that produce?